client/inflight.go

```diff
@@ -0,0 +1,313 @@
+		if !ok {
+			return -1, fmt.Errorf("%v: handler closed", h)
+		}
+		log.Debug().Msgf("%v: borrowed stream id: %v", h, id)
+		return id, nil
+	default:
+		return -1, fmt.Errorf("%v: no stream id available", h)
+	}
+}
+
+func (h *inFlightRequestsHandler) releaseStreamId(id int16) error {
+	if h.isClosed() {
+		return fmt.Errorf("%v: handler closed", h)
+	}
+	select {
+	case h.streamIds <- id:
+		log.Debug().Msgf("%v: released stream id: %v", h, id)
+		return nil
+	default:
+		return fmt.Errorf("%v: stream id %d: release failed", h, id)
+	}
+}
+
+func (h *inFlightRequestsHandler) isClosed() bool {
+	return atomic.LoadInt32(&h.closed) == 1
+}
+
+func (h *inFlightRequestsHandler) setClosed() bool {
+	return atomic.CompareAndSwapInt32(&h.closed, 0, 1)
+}
+
+func (h *inFlightRequestsHandler) close() {
+	if h.setClosed() {
+		log.Trace().Msgf("%v: closing", h)
+		h.inFlightLock.Lock()
+		for streamId, inFlight := range h.inFlight {
+			delete(h.inFlight, streamId)
+			inFlight.close(fmt.Errorf("%v: handler closed", h))
+		}
+		h.inFlightLock.Unlock()
+		streamIds := h.streamIds
+		h.streamIds = nil
+		close(streamIds)
+		log.Trace().Msgf("%v: successfully closed", h)
+	}
+}
+
+type inFlightRequest struct {
+	handlerId     string
+	streamId      int16
+	_incoming     chan *frame.Frame // used internally; will be set to nil on close
+	incoming      chan *frame.Frame // exposed externally; never nil
+	err           error
+	timeout       time.Duration
+	onDone        func(error)
+	ctx           context.Context
+	cancel        context.CancelFunc
+	timeoutCtx    context.Context
+	timeoutCancel context.CancelFunc
+
+	// lock guards the closing of incoming chan and the assignment of err;
+	// required to fulfill the interface contract:
+	// if Incoming() is closed because of an error, Err() must return that error.
+	lock *sync.RWMutex
+}
+
+func (r *inFlightRequest) StreamId() int16 {
+	return r.streamId
+}
+
+func (r *inFlightRequest) Incoming() <-chan *frame.Frame {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	return r.incoming
+}
+
+func (r *inFlightRequest) Err() error {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	return r.err
+}
+
+func newInFlightRequest(
+	handlerId string,
+	streamId int16,
+	maxPending int,
+	timeout time.Duration,
+	onDone func(error),
+) *inFlightRequest {
+	ctx, cancel := context.WithCancel(context.Background())
+	incoming := make(chan *frame.Frame, maxPending)
+	return &inFlightRequest{
+		handlerId: handlerId,
+		streamId:  streamId,
+		_incoming: incoming,
+		incoming:  incoming,
+		timeout:   timeout,
+		onDone:    onDone,
+		ctx:       ctx,
+		cancel:    cancel,
+		lock:      &sync.RWMutex{},
+	}
+}
+
+func (r *inFlightRequest) String() string {
+	return fmt.Sprintf("%v [stream id %d]", r.handlerId, r.streamId)
+}
+
+func (r *inFlightRequest) onFrameReceived(f *frame.Frame) error {
+	select {
+	case r._incoming <- f:
+		done := isLastFrame(f)
+		if done {
+			r.stopTimeout()
+			defer r.onDone(nil)
```

Comment:
this is releasing the stream id after writing to the channel so if the client is not using the managed mode, it might attempt to use the same stream id and it will fail because it hasn't been released yet (I ran into this issue when running the local server tests)

Reply:
This is releasing the stream id when a frame is received, not when a frame is sent. Can you explain a bit more what was the issue?

Reply:
```
result1 := conn.SendAndReceive(firstFrame)
result2 := conn.SendAndReceive(secondFrame)
```

IIUC if both frames have the same stream id then it is possible that the second `Send` is attempted before the stream id is released on the first `Receive`

Reply:
Here is the error:
```
11:49AM ERR CQL client conn [L:127.0.0.1:53364 <-> R:127.0.0.1:9043]: handshake failed error="could not send AUTH RESPONSE: CQL client conn [L:127.0.0.1:53364 <-> R:127.0.0.1:9043]: failed to register in-flight handler for frame {header: {response: false, version: 65, flags: 00000000, stream id: 1, opcode: 15, body length: 0}, body: {tracing id: <nil>, payload: map[], warnings: [], message: AUTH_RESPONSE}}: CQL client conn [L:127.0.0.1:53364 <-> R:127.0.0.1:9043]: [in-flight handler]: stream id already in use: 1"
```

Note that I had to add this change to get the exact error: https://github.com/datastax/go-cassandra-native-protocol/pull/9#discussion_r519748667

Reply:
> IIUC if both frames have the same stream id then it is possible that the second Send is attempted before the stream id is released on the first Receive

If the frames have stream id zero: the stream id will be assigned automatically, and yes the second frame is sent before the first one receives its response, but that's ok since the frames will have distinct assigned stream ids.

If the frames have stream id > 0: it's up to the called to handle stream ids, in this case `firstFrame` and `secondFrame` should not be sent with the same stream id, but with different ones.

I think in your case judging from the error they were both sent with stream id 1: `stream id already in use: 1`.

Try sending them with stream id zero.

Reply:
But the second frame is being sent AFTER the first ´sendAndReceive` has returned, this error is being thrown by the tests I didn't do any changes, I just ran them

Reply:
The handshake method itself uses the stream id 1 so that is an issue but I still think that when `sendAndReceive` returns, the stream id should have already been released so even if we change the handshake method to use `0` we should still consider the current behavior as a bug

Reply:
Otherwise we should just disable "manual" stream ids altogether imo

Reply:
> The handshake method itself uses the stream id 1

Currently it uses zero actually, and it's not configurable. We can add a `streamId` parameter to `InitiateHandshake`, I will do that.

> the second frame is being sent AFTER the first sendAndReceive has returned.

I just tried this and it is working:

```
	f, _ := frame.NewRequestFrame(version, 1, false, nil, &message.Options{}, false);
	_, _ = clientConn.SendAndReceive(f)
	_, _ = clientConn.SendAndReceive(f)
	_, _ = clientConn.SendAndReceive(f)
	_, _ = clientConn.SendAndReceive(f)
```

Reply:
Maybe the bug is not exactly what I'm describing but a bug definitely exists because the local test server tests sometimes fail with the error message that I pasted here in a previous comment.

Reply:
> Currently it uses zero actually, and it's not configurable. We can add a streamId parameter to InitiateHandshake, I will do that.

Are you sure about this? I think this error would not manifest itself in the tests if this was the case.

Isn't this (https://github.com/datastax/go-cassandra-native-protocol/blob/enhanced-client/client/handshake.go#L60) and this (https://github.com/datastax/go-cassandra-native-protocol/blob/enhanced-client/client/handshake.go#L70) using stream id 1?

Reply:
IMO it's a problem of mixing managed and non-managed stream ids, especially because the handshake method would only use managed stream ids until my latest commit. With my latest commit, you now have the ability to pass the stream id, and you would pass zero to get stream id management activated.

In non-managed mode the connection doesn't even care about which stream id you are sending, that's why I'm surprised that sending a frame with non-zero stream ids would cause a "stream id already in use" error. It's necessarily because of a mix of managed and non-managed frames.

Reply:
Ha, you just found the bug :-p I completely forgot these 2 places. It should be fixed now, thanks for spotting this.

Reply:
👍 I still have a feeling that there's something fishy here, I'll spend a few minutes trying to reproduce what I saw with a specific test

Reply:
Thanks for helping me out, it's definitely very hard to get those in-flight requests + timeouts right. I think we are approaching something fully functional though.